### DISCUSSION TITLE
ESP32 bluetooth I2S Philips, TDA1541

### DISCUSSION THREAD

**Discussion by OxF8** [pschatzmann/ESP32-A2DP#404] [Q&A] [answered]
Hello Dear, I'm new in arduino and ESP32, but is there easy way transmit standard Philips I2S to the TDA1541? I try with 
static i2s_config_t i2s_config = {
      .mode = (i2s_mode_t) (I2S_MODE_MASTER | I2S_MODE_TX),
      .sample_rate = 44100, // updated automatically by A2DP
      .bits_per_sample = (i2s_bits_per_sample_t)16,
      .channel_format = I2S_CHANNEL_FMT_RIGHT_LEFT,
      .communication_format = (i2s_comm_format_t) (I2S_COMM_FORMAT_I2S),
      .intr_alloc_flags = 0, // default interrupt priority
      .dma_buf_count = 8,
      .dma_buf_len = 64,
      .use_apll = true,
      .tx_desc_auto_clear = true // avoiding noise in case of data unavailability
  };  
but sound is too noisy.

**Comment by pschatzmann** [comment, marked as answer]
Maybe your DAC does not support I2S_COMM_FORMAT_I2S. 
Did you try the [other formats](https://docs.espressif.com/projects/esp-idf/en/release-v4.4/esp32/api-reference/peripherals/i2s.html?highlight=i2s#_CPPv4N17i2s_comm_format_t19I2S_COMM_FORMAT_I2SE)? e.g. I2S_COMM_FORMAT_STAND_MSB

**Reply by OxF8** [reply]
Sorry, hardware pb - not filtered supply, & missed resistors on the output of the DAC. No way, TDA1541 not to support Philips I2S standard, DAC is more important than protocol :) Thank You very much for this work, one more coffee :) Only strange to me is C+- for embedded. It should be C. I'm in doubts, somebody will want dynamic binding, multiple inheritance etc :)

**Reply by pschatzmann** [reply]
Arduino is C++ so it i quite natural that this library is C++ as  well.  If you want a C version you can use the example provided by Espressif.
For this project it does not make a big difference, but for my [AudioTools](https://github.com/pschatzmann/arduino-audio-tools) it does. I was able to create a quite elegant and powerful audio library. Compare this with the ADF where you need to create 20 times the amount of code to achieve the same thing!

**Reply by OxF8** [reply]
:) Yes, I understand You are workin in environment, & decision is not Your. Anyway, C with its pointer to functions can change dynamically in a way not restricted as in the OOP :) It is most powerful, and... most dangerous :)

**Reply by pschatzmann** [reply]
Just looked at the [price of a module using this DAC](http://www.analogmetric.com/goods.php?id=428): This is expensive!

**Reply by OxF8** [reply]
there is cheaper variant - TDA1541 replacement, with 8xTDA1387 in parallel(https://www.ebay.com/itm/333542898803). Sound from this DAC is different, and if some mods applied(capacitors, RC-shunts and good filtering of the digital fuckery), it sounds really good. Can be attached to tube amplifier also. I will put this in one old Celica(1984), I prepare it for my son, they are listening a bit different music :)
